packages/utils/ts-utils/src/CompilerHost.js

```diff
@@ -1,22 +1,75 @@
 // @flow
-import type {FilePath} from '@parcel/types';
+import type {FileSystem} from '@parcel/fs';
+import type {FilePath, PackageJSON} from '@parcel/types';
+import type {PluginLogger} from '@parcel/logger';
+import typeof TypeScriptModule from 'typescript'; // eslint-disable-line import/no-extraneous-dependencies
 import type {CompilerOptions, SourceFile} from 'typescript';
 import typeof {ScriptTarget} from 'typescript'; // eslint-disable-line import/no-extraneous-dependencies
+
 import path from 'path';
 import {FSHost} from './FSHost';
 
 export class CompilerHost extends FSHost {
   outputCode: ?string;
   outputMap: ?string;
+  logger: PluginLogger;
+  // workaround for https://github.com/microsoft/TypeScript/issues/39547
+  redirectTypes: Map<FilePath, FilePath> = new Map();
+
+  constructor(fs: FileSystem, ts: TypeScriptModule, logger: PluginLogger) {
+    super(fs, ts);
+    this.logger = logger;
+  }
+
+  readFile(filePath: FilePath): void | string {
+    let contents = super.readFile(filePath);
+    if (contents && path.basename(filePath) === 'package.json') {
+      let json: PackageJSON = JSON.parse(contents);
+      if (
+        json.types != null &&
+        json.source != null &&
+        !super.fileExists(
+          path.posix.join(path.posix.dirname(filePath), json.types),
+        )
+      ) {
+        let source = path.posix.join(path.posix.dirname(filePath), json.source);
+        let fakeTypes =
+          source.slice(0, -path.posix.extname(source).length) + '.d.ts';
+        this.redirectTypes.set(fakeTypes, source);
```

Comment:
Lol, so TS requires it to be a `.d.ts` file but doesn't validate that when parsing? 🤦

Reply:
Something along those lines, I don't quite remember.